### ISSUE TITLE
looking for read_structure column in samplesheet?

### ISSUE THREAD

**Issue by MikeHala** [nf-core/fastquorum#10]
### Description of the bug

Thank you for providing the fastquorum pipeline!

I have tried to install it and perform the test run with the command:
`nextflow run nf-core/fastquorum -profile test,singularity --outdir my_test -r dev` 

The first issue was:
```
ERROR: Validation of pipeline parameters failed!
* Missing required parameter: --ref_fasta
```
 I have provided a random (human) fasta file, but when I tried again
`./nextflow run nf-core/fastquorum -profile test,singularity --outdir mike_test --ref_fasta "Homo_sapiens_assembly38.fasta" -r dev`

I ran into the second issue

```
N E X T F L O W  ~  version 22.10.0
Launching `https://github.com/nf-core/fastquorum` [desperate_galileo] DSL2 - revision: e89ab00e02 [dev]

Core Nextflow options
  revision                  : dev
  runName                   : desperate_galileo
  containerEngine           : singularity
  launchDir                 : /gpfs/igmmfs01/eddie/bioinfsvice/fastquorum_test
  workDir                   : /gpfs/igmmfs01/eddie/bioinfsvice/fastquorum_test/work
  projectDir                : /home/mhalache/.nextflow/assets/nf-core/fastquorum
  userName                  : mhalache
  profile                   : test,singularity
  configFiles               : /home/mhalache/.nextflow/assets/nf-core/fastquorum/nextflow.config

Input/output options
  input                     : https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/samplesheet/samplesheet_test_illumina_amplicon.csv
  ref_fasta                 : /exports/igmm/eddie/BioinformaticsResources/igenomes/Homo_sapiens/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38.fasta
  outdir                    : mike_test

Main options
  duplex_seq                : true

Read grouping options
  groupreadsbyumi_edits     : 1

Consensus reads options
  call_min_baseq            : 10

Consensus filtering options
  filter_min_baseq          : 10
  filter_max_base_error_rate: 0.1

Reference genome options
  genome                    : R64-1-1
  fasta                     : s3://ngi-igenomes/igenomes/Saccharomyces_cerevisiae/Ensembl/R64-1-1/Sequence/WholeGenomeFasta/genome.fa

Institutional config options
  config_profile_name       : Test profile
  config_profile_description: Minimal test dataset to check pipeline function

Max job request options
  max_cpus                  : 2
  max_memory                : 6.GB
  max_time                  : 6.h

!! Only displaying parameters that differ from the pipeline defaults !!
------------------------------------------------------
If you use nf-core/fastquorum for your analysis please cite:

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x

* Software dependencies
  https://github.com/nf-core/fastquorum/blob/master/CITATIONS.md
------------------------------------------------------
executor >  local (1)
[16/159a3d] process > NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK (samplesheet_test_illumina_amplicon.csv) [100%] 1 of 1, failed: 1 ✘
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:FASTQC                                                                 -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:CUSTOM_DUMPSOFTWAREVERSIONS                                            -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:FASTQTOBAM                                                             -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:ALIGN_RAW_BAM                                                          -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:GROUPREADSBYUMI                                                        -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:CALLDDUPLEXCONSENSUSREADS                                              -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:COLLECTDUPLEXSEQMETRICS                                                -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:ALIGN_CONSENSUS_BAM                                                    -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:FILTERCONSENSUSREADS                                                   -
[-        ] process > NFCORE_FASTQUORUM:FASTQUORUM:MULTIQC                                                                [  0%] 0 of 1
Execution cancelled -- Finishing pending tasks before exit
-[nf-core/fastquorum] Pipeline completed with errors-
Error executing process > 'NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK (samplesheet_test_illumina_amplicon.csv)'

Caused by:
  Process `NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK (samplesheet_test_illumina_amplicon.csv)` terminated with an error exit status (1)

Command executed:

  check_samplesheet.py \
      samplesheet_test_illumina_amplicon.csv \
      samplesheet.valid.csv

  cat <<-END_VERSIONS > versions.yml
  "NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK":
      python: $(python --version | sed 's/Python //g')
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  INFO:    Convert SIF file to sandbox...
  WARNING: Skipping mount /exports/eddie3_apps_local/apps/SL7/singularity/3.5.3/var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
  [CRITICAL] The sample sheet **must** contain the column headers: fastq_2, fastq_1, sample, read_structure.
  INFO:    Cleaning up image...

Work dir:
  /gpfs/igmmfs01/eddie/bioinfsvice/fastquorum_test/work/16/159a3d29d803243ed1b1e66950e9c7

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
```

Any advice will be greatly appreciated!

Best,
Mike

**Comment by nh13** [comment]
Can you post the sample sheet?  Did you make sure it has teh correct columns:

> The sample sheet **must** contain the column headers: fastq_2, fastq_1, sample, read_structure

**Comment by MikeHala** [comment]
Hi,

Thank you for your reply and apologies for the delay (took me some time to obtain the read_structure info from the sequencing provider).

I am still struggling with parsing the sample sheet

```
Core Nextflow options
  revision                  : dev
  runName                   : fabulous_majorana
  containerEngine           : singularity
  launchDir                 : /gpfs/igmmfs01/eddie/ETR_BREAST_CANCER
  workDir                   : /gpfs/igmmfs01/eddie/ETR_BREAST_CANCER/work
  projectDir                : /exports/igmm/eddie/ETR_BREAST_CANCER/.nextflow/assets/nf-core/fastquorum
  userName                  : mhalache
  profile                   : singularity
  configFiles               : /exports/igmm/eddie/ETR_BREAST_CANCER/.nextflow/assets/nf-core/fastquorum/nextflow.config

Input/output options
  input                     : /exports/igmm/eddie/ETR_BREAST_CANCER/test_samples.csv
  ref_fasta                 : /exports/igmm/eddie/BioinformaticsResources/igenomes/Homo_sapiens/GATK/GRCh38/Sequence/WholeGenomeFasta/Homo_sapiens_assembly38.fasta
  outdir                    : test_results
  email                     : [xxx]
  multiqc_title             : test

Read grouping options
  groupreadsbyumi_edits     : 1

Consensus reads options
  call_min_baseq            : 10

Consensus filtering options
  filter_min_baseq          : 10
  filter_max_base_error_rate: 0.1

Execution cancelled -- Finishing pending tasks before exit
-[nf-core/fastquorum] Sent summary e-mail to [xxx] (sendmail)-
-[nf-core/fastquorum] Pipeline completed with errors-
Error executing process > 'NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK (test_samples.csv)'

Caused by:
  Process `NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK (test_samples.csv)` terminated with an error exit status (1)

Command executed:

  check_samplesheet.py \
      test_samples.csv \
      samplesheet.valid.csv

  cat <<-END_VERSIONS > versions.yml
  "NFCORE_FASTQUORUM:FASTQUORUM:INPUT_CHECK:SAMPLESHEET_CHECK":
      python: $(python --version | sed 's/Python //g')
  END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  INFO:    Convert SIF file to sandbox...
  WARNING: Skipping mount /exports/eddie3_apps_local/apps/SL7/singularity/3.5.3/var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
  [CRITICAL] The given sample sheet does not appear to contain a header.
  INFO:    Cleaning up image...

Work dir:
  /gpfs/igmmfs01/eddie/ETR_BREAST_CANCER/work/7e/e69ea9ab8ed3bd2485e60edb8494bd

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line
```
and here is the content of test_samples.csv

```
sample,fastq_1,fastq_2,read_structure
NA12878,raw_fastq/NA12878_S95_L001_R1_001.fastq.gz,raw_fastq/NA12878_S95_L001_R2_001.fastq.gz,151T12M11S128T
OncoSpan_gDNA,raw_fastq/OncoSpan_gDNA_S96_L001_R1_001.fastq.gz,raw_fastq/OncoSpan_gDNA_S96_L001_R2_001.fastq.gz,151T12M11S128T
Ref2_Quant,raw_fastq/Ref2_Quant_S31_L001_R1_001.fastq.gz,raw_fastq/Ref2_Quant_S31_L001_R2_001.fastq.gz,151T12M11S128T
Ref3_Struct,raw_fastq/Ref3_Struct_S57_L001_R1_001.fastq.gz,raw_fastq/Ref3_Struct_S57_L001_R2_001.fastq.gz,151T12M11S128T
RS102_Local,raw_fastq/RS102_Local_S36_L001_R1_001.fastq.gz,raw_fastq/RS102_Local_S36_L001_R2_001.fastq.gz,151T12M11S128T
```

I have also tried a tab-separated version of the sample sheet, as well as a version in which the order is "old_test_samples.csv" (as per your comment above)...

What am I doing wrong?

Thanks,
Mike

**Comment by MikeHala** [comment]
Hi again,

I think I found a solution to the problem. The problematic sample sheet was generated using the nano text editor; when I entered the information in MS Word and then copy/pasted it to nano it worked (I suppose there is some issue with the newline character?).

Also, had to modify the read_structure values, to separate (by space) the info for R1 and R2, i.e.
```
sample,fastq_1,fastq_2,read_structure
NA12878,raw_fastq/NA12878_S95_L001_R1_001.fastq.gz,raw_fastq/NA12878_S95_L001_R2_001.fastq.gz,151T 12M11S128T
....
```

I am closing this issue, 
sorry and thanks.
Mike
